### PR TITLE
Power control (and some refactoring)

### PR DESCRIPTION
**Pull request by kbembedded**
TODO: Need to add options to cmdline to set specific power states at startup

Also, @FryDay, would you be interested in adding some more graphics like you did for the LED? Basically a battery indicator like you'd see on a smart device: white battery outline, two bars on the inside, both for high, one for low, none (and maybe change the battery outline to red?) for depleted. And some kind of AC/external power input indicator. I'm not sure the best way to showcase that. If you don't have any ideas I can dig around some more. Also, feel free to comment on changes, it looks like you arn't a part of this project normally so I can't request a review.

### CODE REVIEW

File: src/msemu.c [hunk 1]
```diff
@@ -115,7 +115,8 @@ void ms_power_on_reset(ms_ctx *ms)
 
 	lcd_init(&ms->lcd_dat1bit, &ms->lcd_datRGBA8888, &ms->lcd_cas);
 	ram_init(&ms->ram, NULL);
-	io_init(&ms->io);
+	//io_init(&ms->io);
```
Comment (kbembedded) [inline src/msemu.c]: Whoops, didn't mean to leave this commented out.

File: src/msemu.c [hunk 2]
```diff
@@ -382,7 +384,19 @@ Z80EX_BYTE z80ex_pread (
 	   * states for testing?
 	   */
 	  case MISC9:
-		ret = (uint8_t)0xE0 | ((~ms->power_button & 1) << 4);
+		tmp = 0;
+		if (ms->ac_status) tmp |= MISC9_AC_GOOD;
+		// XXX: Make the 4 a define
```
Comment (kbembedded) [inline src/msemu.c:391]: Forgot about this...

File: src/msemu.c [hunk 3]
```diff
@@ -451,7 +465,7 @@ void z80ex_pwrite (
 	Z80EX_BYTE val,
 	void *user_data)
 {
-
+	uint8_t tmp;
```
Comment (kbembedded) [inline src/msemu.c:470]: This can go

File: src/msemu.c [hunk 4]
```diff
@@ -568,6 +582,55 @@ int process_interrupts (ms_ctx* ms)
 	return 0;
 }
 
+/* Enable, disable, or toggle AC adapter status
+ *
+ * Writes to the relevant IO port
```
Comment (kbembedded) [inline src/msemu.c:589]: Doesn't actually write to IO, fix comment

File: src/msemu.c [hunk 5]
```diff
@@ -568,6 +582,55 @@ int process_interrupts (ms_ctx* ms)
...
+		break;
+	case AC_FAIL:
+		printf("AC fail\n");
+		break;
+	}
+}
+
+/* Set battery voltage status, high, low, depleted
+ *
+ * Writes to the relevant IO port
```
Comment (kbembedded) [inline src/msemu.c:611]: Fix comment, doesn't write to IO port

File: src/msemu.c [hunk 6]
```diff
@@ -583,7 +646,8 @@ int ms_init(ms_ctx* ms, ms_opts* options)
 	srand((unsigned int)time(NULL));
 
 	ms->interrupt_mask = 0;
-	ms->power_button = 0;
+	// XXX: Should this be abstracted? Seems silly to do so.
```
Comment (kbembedded) [inline src/msemu.c:651]: Figure out what to do with this

File: src/msemu.c [hunk 7]
```diff
@@ -600,6 +664,8 @@ int ms_init(ms_ctx* ms, ms_opts* options)
 	if (lcd_init(&ms->lcd_dat1bit, &ms->lcd_datRGBA8888, &ms->lcd_cas))
 		return MS_ERR;
 	if (io_init(&ms->io)) return MS_ERR;
+	ms_power_ac_set_status(ms, AC_GOOD);
```
Comment (kbembedded) [inline src/msemu.c:668]: Change these to be evaluated from command line, default to AC_GOOD and BATT_HIGH though

File: src/mem.c [hunk 8]
```diff
@@ -290,32 +290,32 @@ int cf_init(uint8_t **cf_buf, ms_opts *options)
...
         }
 
 	return MS_OK;
 }
 
-int cf_deinit(uint8_t **cf_buf, ms_opts *options)
+int cf_deinit(ms_ctx *ms, ms_opts *options)
 {
-	assert(*cf_buf != NULL);
+	assert(ms->cf != NULL);
```
Comment (kbembedded) [inline src/mem.c:305]: Remove this, waste of cpu time

**Comment by FryDay** [comment]
@kbembedded I'd be happy to work on some more graphics. I'll let you know what kind of ideas I come up with.

**Comment by FryDay** [comment]
#64 implements the new indicators for the battery and ac status.

**Comment by kbembedded** [comment]
This should all be good to go!

**Pull request merged**